pageserver/src/layered_repository.rs

```diff
@@ -1976,7 +1986,16 @@ impl LayeredTimeline {
         drop(layers);
 
         // Only compact if enough layers have accumulated.
-        if level0_deltas.is_empty() || level0_deltas.len() < self.get_compaction_threshold() {
+        // Also force compaction if L0 delta layers contains a lot of dup;icatd keys.
+        // As far as compaction can not split key ito two layers, at may cause generation of large layer files
+        // (up to compaction_threshold*checkpoint_distance).
+        if level0_deltas.is_empty()
+            || (level0_deltas.len() < self.get_compaction_threshold()
+                && level0_deltas
+                    .iter()
+                    .fold(0u64, |acc, l| acc + l.get_max_key_range().unwrap())
+                    < target_file_size)
+        {
```

Comment:
Hmm, so the idea is to prevent too many L0 files with the same keys from accumulating, so that the logic below doesn't create a too large file. I don't know; that probably works OK in practice. However, you could still end up in the same situation if the compaction isn't triggered in time. Maybe the server is busy so it didn't get around to do compaction fast enough. Or maybe it was restored from S3 after a crash, and the L1 files had not been uploaded to S3 yet, or something.

I would prefer a two-stage approach in the loop below instead. It would seem more robust.

Reply:
Yes, the idea was to limit number of compacted L0 layers. In my case maximal layer size can be up to 2*`target_file_size`  and arbitrary small "tail" layer. Right now this check is used only to trigger compaction and as you   have noticed actual  number of compacted layers can be larger. But is quite trivial to strictly limit number of compacted layers, i.e. when `level0_deltas` is constructed. 

There are two main advantages of this approach:
1. It is simple (just need to add `get_max_key_range` function and perform simple aggregation)
2. Minimize number of small delta layers
3. Do not require two passes

Disadvantages:
1 Layer can be up to 2 times bigger than specified by `target_file_size`
2. By limiting number of compacted layers we make compaction less efficient.
3. I am calculated maximal key range for each layer, but them can refer to different keys. So estimation is not so precise.

It is also not so difficult to implement alternative approach: actually we even do not need to make two passes.
It is enough to calculate size of next key. Right now it is not so difficult to do and it is not require extra pass because `DeltaValueIter` contains `all_offsets: Vec<(DeltaKey, BlobRef)>`.
So we cn check that `current_layer_size + next_key_size  < target_file_size`.
In this case there will be no layers larger than `target_file_size`. But there cab be multiple small layers (and I do no see any way how we can avoid creation of them even with the smartest repartition planner)

So, what do you think? Should I just sightly modify my current implementation by limiting size of `level0_deltas`.
Or try to implement alternative approach? It is not cear to me how to propagate calculation of key range through `kmerge_by`. Certainly I can just add method to `DeltaValueIter::get_key_range(Key)`. But in this case binary search in `all_offsets` array will be needed. I afraid that it will have negative ipact on performance.

Reply:
> It is also not so difficult to implement alternative approach: actually we even do not need to make two passes. It is enough to calculate size of next key. Right now it is not so difficult to do and it is not require extra pass because `DeltaValueIter` contains `all_offsets: Vec<(DeltaKey, BlobRef)>`. So we cn check that `current_layer_size + next_key_size < target_file_size`. In this case there will be no layers larger than `target_file_size`. But there cab be multiple small layers (and I do no see any way how we can avoid creation of them even with the smartest repartition planner)

Works for me. I'm also not too worried about having an occasional small layer.